### PR TITLE
Add Diagnostics on Rule Create and Delete

### PR DESCRIPTION
**Pull request by molly-moen**
Add call to diagnostics service to log event on rule creation and deletion. On both rule create and delete, a Rule_Created/Rule_Deleted event will be emitted as well as a Rule_Count metric which will report the updated count of non-deleted rules. These metrics will only be emitted on successful operations. The diagnostics url is provided as an environment variable, but if the url is not provided, the microservice will still run as before, and will not attempt to log to diagnostics.

### CODE REVIEW

File: device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java [hunk 1]
```diff
@@ -456,4 +461,56 @@ private RuleServiceModel getServiceModelFromJson(JsonNode response) {
                     "Could not parse data from Key Value Storage"));
         }
     }
-}
+
+    private CompletionStage<Integer> getRuleCountAsync() {
+        return this.prepareRequest(null)
```
Comment (zhang-hua) [inline device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: nit: prepareRequest don't have parameter.

Comment (molly-moen) [reply device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/Rules.java]: it has a parameter of id

File: device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+public class DiagnosticsClient implements IDiagnosticsClient {
+
+    private final WSClient wsClient;
+    private final String diagnosticsEndpointUrl;
+    private final int diagnosticsMaxLogRetries;
+    private final boolean canWriteToDiagnostics;
+
+    private static final Logger.ALogger log = Logger.of(Rules.class);
+    private static final int OK = 200;
```
Comment (zhang-hua) [inline device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java]: nit: we can use it from Http lib without redefine it.

File: device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                        log.error("Error logging to diagnostics service {}",
+                            error.getMessage());
+                        throw new CompletionException(
+                            new ExternalDependencyException(result.getStatusText()));
+                    }
+                    return true;
+                });
+    }
+
+    private WSRequest prepareRequest() {
```
Comment (zhang-hua) [inline device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java]: missing 'id' parameter here?

Comment (molly-moen) [reply device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java]: no we don't need an id for the request to diagnosticsevents

**Pull request merged**